test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,391 @@
+// Copyright 2024 Antrea Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	crdv1beta1 "antrea.io/antrea/pkg/apis/crd/v1beta1"
+	agentconfig "antrea.io/antrea/pkg/config/agent"
+	"antrea.io/antrea/test/e2e/utils"
+)
+
+/*
+1) Create the KIND cluster.
+2) Once the cluster is up , create a service for custom DNS.
+3) Configure and update Antrea configmap with above service's IP.
+
+5) Create agnHost 2 pods (deployment?).
+6) Get IP of one of the pods of agnHost.
+7) create and configure the custom CoreDNS configMap with the IP received above.
+8) Create a custom DNS pod and configure it to use above configMap with agnHost pod's IP.
+9) Create and apply antrea FQDN policy.
+10) Deploy antrea-toolbox.
+
+11) curl the FQDN from within toolbox.
+12) imitate caching the IP belonging to above FQDN resolution by keeping it in a variable.
+13) change the existing IP in dns configmap with the other agnHost pod's IP.
+14) wait for new IP to get updated in configMap and let the changes be reflected in dns pod.
+15) curl the FQDN again with old IP , simulating usage of cache -- and it must fail with no connectivity.
+*/
+
+// agnHostPodIpv4Addresses type stores the IPv4 addresses of the 2 agnHost pods created for the test.
+type agnHostPodIpv4Addresses struct {
+	pod1Ipv4 string
+	pod2Ipv4 string
+}
+
+func TestFQDNPolicyWithCachedDNS(t *testing.T) {
+	const testFQDN = "fqdn-test-pod.lfx.test"
+
+	skipIfAntreaPolicyDisabled(t)
+	skipIfNotIPv4Cluster(t)
+	skipIfIPv6Cluster(t)
+
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	dnsServiceIP := setCustomDnsServerAddressInAntrea(t, data, false)
+	defer setCustomDnsServerAddressInAntrea(t, data, true)
```

Comment:
It is quite clumsy to create the K8s Service (custom DNS) as part of this function.
The code should look like this:
```
// setup custom DNS
// set DNS server for Antrea
// defer restore DNS server for Antrea
```

It seems we could also create the DNS Pod prior to setting the DNS server address for Antrea? There isn't really a circular dependency issue here since Antrea is already running.